torch/distributed/device_mesh.py

```diff
@@ -488,6 +341,194 @@ def __init__(
                     rank_coords[0].tolist() if rank_coords.size(0) > 0 else None
                 )
 
+        def _init_process_group(
+            self,
+            layout: _Layout,
+            dim: int,
+            name: Optional[str] = None,
+            backend_override: tuple[Optional[str], Optional[C10dBackend.Options]] = (
+                None,
+                None,
+            ),
+            array_mesh: Optional[torch.Tensor] = None,
+        ) -> Optional[ProcessGroup]:
+            """
+            Creates a process group backend for a given layout and dimension.
+
+            Args:
+                layout (_Layout): The layout object representing the communication pattern
+                dim (int): The dimension index in the device mesh
+                name (Optional[str]): Optional name for the mesh dimension
+                backend_override (tuple[Optional[str], Optional[C10dBackend.Options]]):
+                    Optional backend type and options to override default settings
+                array_mesh (Optional[torch.Tensor]): Optional array mesh tensor to use for subgroup creation when the mesh
+                    is not CuTe expressible
+            """
+            default_group = _get_default_group()
+            if layout == _Layout((get_world_size(),), (1,)) and backend_override == (
+                None,
+                None,
+            ):
+                # Append the default pg to the first dim groups only if the default pg is compatible with `self.device_type`.
+                # Otherwise, create new pg.
+                ranks = list(range(get_world_size()))
+                return (
+                    new_group(
+                        backend="cpu:gloo,cuda:nccl",
+                        ranks=ranks,
+                        group_desc="mesh_default",
+                    )
+                    if torch.cuda.is_available()
+                    and get_backend(default_group) == "gloo"
+                    else default_group
+                )
+            else:
+                # Generate the pg_ranks_by_dim for pg_creation from the layout.
+                # When Mesh ranks are not representible by CuTe layout, we want to directly
+                # use the mesh ranks to create PGs.
+                pg_ranks_by_dim = (
```

Comment:
based on this comment https://github.com/pytorch/pytorch/pull/161106#issuecomment-3224801940
i think we should normalize the way we get ranks for new_group calls.  We should have one way of doing it that we always use, and it should be to take the mesh_tensor and apply the layout to it.

The mesh_tensor would default to `range(n)` if the user did not provide a mesh_tensor, but if they do provide one it lets them fully customize the device ordering.

Wdyt?